gap/congruences/congpairs.gi

```diff
@@ -404,22 +404,24 @@ InstallMethod(EquivalenceClasses,
 "for an fp semigroup congruence",
 [IsFpSemigroupCongruence],
 function(cong)
-  local part, enum, reps, classes, next, i;
-
-  part := CongruenceByGeneratingPairsPartition(cong);
+  local enum, nrclasses, classes, class_added, i, word, id;
   enum := Enumerator(Range(cong));
-  reps := List(part, x -> enum[x[1]]);
-
-  classes := EmptyPlist(Length(reps));
-  next := 1;
-
-  for i in [1 .. Length(reps)] do
-    classes[next] := EquivalenceClassOfElementNC(cong, reps[i]);
-    SetCongruenceClassByGeneratingPairsCosetId(classes[next], i);
-    SetSize(classes[next], Length(part[i]));
+  nrclasses := NrEquivalenceClasses(cong);
 
-    next := next + 1;
-  od;
+  # Add each class when we find a representative for it
+  classes := EmptyPlist(nrclasses);
+  class_added := BlistList([1 .. nrclasses], []);
+  i := 0;
+  repeat
+    i := i + 1;
+    word := SEMIGROUPS.ExtRepObjToWord(ExtRepOfObj(enum[i]));
+    id := CONG_PAIRS_ELM_COSET_ID(cong, word);
+    if not class_added[id] then
+      classes[id] := EquivalenceClassOfElementNC(cong, enum[i]);
+      SetCongruenceClassByGeneratingPairsCosetId(classes[id], id);
+      class_added[id] := true;
+    fi;
+  until not false in class_added;
```

Comment:
Presumably this requires a linear search every time the loop happens. How about: have a counter, it starts at `0`, you add `1` to it each time you hit the `if` block of code, and then the `until` line would be `until count = nrclasses;`, or whatever.